File: SOLID-Java/src/main/java/dependencyinversion/before/DatabaseHandler.java
package dependencyinversion.before;

import dependencyinversion.after.Database;

public class DatabaseHandler {
	
	public void connect(Object db) throws Exception {
		if(db instanceof MySQLDatabase) {
			MySQLDatabase mySQL = (MySQLDatabase) db;
			mySQL.connect();
		} else if(db instanceof OracleDatabase) {
			OracleDatabase oracle = (OracleDatabase) db;
			oracle.connect();
		} else {
			throw new Exception("DB not supported");
		}
	}
	
	public void disconnect(Database db) throws Exception {
		if(db instanceof MySQLDatabase) {
			MySQLDatabase mySQL = (MySQLDatabase) db;
			mySQL.disconnect();
		} else if(db instanceof OracleDatabase) {
			OracleDatabase oracle = (OracleDatabase) db;
			oracle.disconnect();
		} else {
			throw new Exception("DB not supported");
		}
	}
}

File: SOLID-Scala/src/main/scala/dependencyinversion/before/DatabaseHandler.scala
package dependencyinversion.before

class DatabaseHandler {
  
  def connect(db: Object) = {
    db match {
      case db: MySQLDatabase => db.connect
      case db: OracleDatabase => db.connect
      case _ => throw new Exception("DB not supported")
    }
	}
	
	def disconnect(db: Object) = {
		db match {
      case db: MySQLDatabase => db.disconnect
      case db: OracleDatabase => db.disconnect
      case _ => throw new Exception("DB not supported")
    }
	}
}